compiler/rustc_expand/src/expand.rs

```diff
@@ -162,77 +185,114 @@ macro_rules! ast_fragments {
     }
 }
 
+fn elems_to_string<T>(elems: &SmallVec<[T; 1]>, f: impl Fn(&T) -> String) -> String {
+    let mut s = String::new();
+    for (i, elem) in elems.iter().enumerate() {
+        if i > 0 {
+            s.push('\n');
+        }
+        s.push_str(&f(elem));
+    }
+    s
+}
+
+fn unreachable_to_string<T>(_: &T) -> String {
+    unreachable!()
+}
+
 ast_fragments! {
-    Expr(P<ast::Expr>) { "expression"; one fn visit_expr; fn visit_expr; fn make_expr; }
-    Pat(P<ast::Pat>) { "pattern"; one fn visit_pat; fn visit_pat; fn make_pat; }
-    Ty(P<ast::Ty>) { "type"; one fn visit_ty; fn visit_ty; fn make_ty; }
+    Expr(P<ast::Expr>) {
+        "expression";
+        one fn visit_expr; fn visit_expr; fn pprust::expr_to_string;
```

Comment:
```suggestion
        one fn visit_expr; fn visit_expr; pprust::expr_to_string;
```
`fn` is used because this macro generates definitions for all these functions and it's convenient if the definitions can be found by searching for `fn foo`, but this macro doesn't define `pprust::expr_to_string`.

Style nit: thus it would look better on the same line with `"expression";`.

Reply:
> fn is used because this macro generates definitions for all these functions

No, it doesn't, it just calls them. We could get rid of all the `fn`s, they aren't necessary.

As for the position, I put these three on the same line because they're the three that get grouped with the `one` or the `many`.

Reply:
I tried removing them, but I get this error:
```
error: local ambiguity when calling macro `ast_fragments`: multiple parsing options: built-in NTs ident ('make_ast') or 1 other option.
   --> compiler/rustc_expand/src/expand.rs:206:9
    |
206 |         one visit_expr; visit_expr; pprust::expr_to_string;
    |         ^^^
```
Let's just keep them.